src/components/gamespec/Board.js

```diff
@@ -5,18 +5,54 @@ import ItemTypes from './ItemTypes';
 import { Layer, Stage } from 'react-konva';
 
 import CanvasImage from './CanvasImage';
+import { piecesRef } from '../../firebase';
 
 const boxTarget = {
   drop(props, monitor, component) {
       let offset = monitor.getClientOffset(),
           item = monitor.getItem();
 
       let items = props.getItems();
+      let pieces = props.getPieces();
       let rect = component.refs.stage.getStage().getContainer().getBoundingClientRect();
       offset.x = offset.x - rect.left;
       offset.y = offset.y - rect.top;
+      let images = [];
+      items.forEach((element, index) =>{
+        let x = element.offset.x - offset.x
+        let y = element.offset.y - offset.y
+        if (Math.pow(x,2) < Math.pow(element.image.width,2) && Math.pow(y,2) < Math.pow(element.image.height,2)) {
+          pieces.forEach((piece, index) =>{
+            if(piece.childKey === element.childKey){
```

Comment:
Also check if the piece was already on the board i.e. was present in some piece. Easy way to do this would be to `Piece.js` file in which you should item to the piece in which the image is present and the index of the image in that piece. Also if the piece gets empty delete it from firebase.

Reply:
I think you are mostly getting error due to that. You are also not removing the image from piece in which it was already present. So if it was already present in a piece, you should have two firebase sets in total

Reply:
Is there any possibility that one image can belong to many pieces?

Reply:
Yes, that is an obvious possibility. Consider the game of Go, you can drag same black piece to board multiple times creating new pieces each time. So, in all of these new pieces created black piece is the same image that we dragged again and again on the board.

Reply:
Do you mind explaining again how our error generates cause the first paragraph is hard to understand?

Reply:
Were you able to figure out this?

Reply:
We've solved the problems and pushed the executable code. Could you check the latest commit and examine whether it can be merged?